rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,350 @@
+
+![Local to Cloud](https://app.lucidchart.com/publicSegments/view/80361b71-5317-478b-921e-0ead5152c865/image.png)
+**Figure 3:** Architecture of a user uploading from their local machine to our cloud.
+
+#### 1. Upload File
+
+Using the browser App, the user select a file to upload and send a request to upload the file.
+
+#### 2. Request Token or Presigned URL
+
+Depending on the size of the file a different response will be returned. If the file is >5GB
+
+#### 3. Upload Directly S3
+
+Using the access token and the AWS JS SDK we will upload the file directly to S3. We will calculate the hash of each chunk
+locally and add the hash to [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) 
+in the Put Object request. This will verify the integrity of each chunk during the upload. If the hash does not match
+the chunk will be rejected.
+
+##### User Upload Limitation
+
+S3 is our CSP, and is where a user's uploaded files will eventually end up. S3 will allow small files to be upload without
+using multipart upload. This allows us to generate a single presigned URL that the user can use to securely upload their
+files.
+
+For large files, AWS S3 requires that the files be uploaded using their multipart upload process. This requires multiple
+requests to S3, and cannot be completed with a single presigned URL. In this case, a temporary security token is generated
+for the user. Token will grants the bearer directs access to the S3 bucket. This allows the browser application to use
+the AWS JS SDK to upload a large file in multiple parts with ease. It important to limit the permissions of the token
+to only allow it upload the expected file. There is also a maximum time to live of one hours for the token, therefore if
+the upload is not complete before then, a new token must be issued.
+
+### Access Token
+
+The access token generated and provided to the user grants direct access to the S3 bucket for uploading. This will allow
+the browser APP to use the AWS JS SDK to perform the multipart uploads to S3 for large files. A backend sever will be used
+to generate the token and additional sessions policies will be applied to the token to further restricts access
+
+### Cleanup Multipart Uploads
+
+In some cases a multipart upload may fail to complete and be left partial uploaded to s3. AWS will retain these partial
+uploads and charge for the storage. These charges can be mitigated by applying an S3 bucket life cycle rule to clean up
+incomplete multipart uploads, see [link](https://aws.amazon.com/blogs/aws/s3-lifecycle-management-update-support-for-multipart-uploads-and-delete-markers/).
+
+### Completing an Upload
+
+Once the file has been uploaded to our bucket additional process may take place. The bucket can be configure to emit events
+when a file has been uploaded that can trigger additional processing. The additional processing is outside the scope of this
+RFC.
+
+### Test plan
```

Comment:
Random comment - How does the design address simple filtering/validation by file format? For example, we do not want to allow upload of a dataset.mtx.

Reply:
This was outside the scope of the ticket, but I image it could be added to the backend before downloading for cloud to cloud or in the frontend before doing a local upload.

Reply:
If validation means to check the file extension is correct, then this can happen on the backend before the file is queued for download or an access token is generated.